lib/api/objectDelete.js

```diff
@@ -24,48 +28,124 @@ export default function objectDelete(authInfo, request, log, cb) {
-                    error: err,
-                    method: 'metadataValidateAuthorization',
-                });
-                return cb(err, corsHeaders);
-            }
+
+    const canonicalID = authInfo.getCanonicalID();
+    let bucketMD = undefined;
+    let objectMD = undefined;
+    let corsHeaders = undefined;
+    let removeDeleteMarker = false;
+    let deleteOptions = undefined;
+    return async.waterfall([
+        callback => services.metadataValidateAuthorization(valParams, callback),
+        (bucket, objMD, callback) => {
+            corsHeaders = collectCorsHeaders(request.headers.origin,
+                    request.method, bucket);
+            bucketMD = bucket;
+            objectMD = objMD;
+            const versioning = bucketMD.getVersioningConfiguration();
+            // TODO adhere to the behaviour of AWS: create a delete marker
+            // even if the deleting object does not exist
             if (!objMD) {
-                return cb(errors.NoSuchKey, corsHeaders);
+                if (versioning) {
+                    if (reqVersionId) {
+                        corsHeaders['x-amz-version-id'] =
+                            VID.encrypt(reqVersionId);
+                        return cb(null, corsHeaders, null);
+                    }
+                } else {
+                    return callback(errors.NoSuchKey);
+                }
+            }
+            if (objMD) {
+                const headerValResult = validateHeaders(objMD, request.headers);
+                if (headerValResult.error) {
+                    return callback(headerValResult.error);
+                }
+                if (objMD['content-length']) {
+                    log.end().addDefaultFields({
+                        contentLength: objMD['content-length'],
+                    });
+                }
             }
-            const headerValResult = validateHeaders(objMD, request.headers);
-            if (headerValResult.error) {
-                return cb(headerValResult.error, corsHeaders);
+            return callback();
+        },
+        callback => services.preprocessingVersioningDelete(bucketName,
+            bucketMD, objectKey, objectMD, reqVersionId, log, callback),
+        (options, callback) => {
+            if (options && options.deleteData) {
+                // delete object
+                deleteOptions = options;
+                return callback(null, options);
             }
-            if (objMD['content-length']) {
-                log.end().addDefaultFields({
-                    contentLength: objMD['content-length'],
-                });
+            // putting a new delete marker
+            if (bucketMD.hasDeletedFlag() &&
+                    canonicalID !== bucketMD.getOwner()) {
+                log.trace('deleted flag on bucket and request ' +
+                        'from non-owner account');
+                return callback(errors.NoSuchBucket);
```

Comment:
Since the bucket have deleting flag, and the requester is not the bucket owner, does that make sense to return AccessDenied ? since the bucket is supposed to be deleted already. We copy this logic from the objectPut :)